devdoc/twodarray_requirements.md

```diff
@@ -0,0 +1,167 @@
+# `two_d_array` requirements
+============================
+
+## Overview
+
+`TWODARRAY` is a module provides a 2D array with variable size of rows and columns. Each row has the same size to ensure easier access. The first dimension of array is resizable, it contains the pointer to the second dimension of array. The pointer is initialized to NULL and points to the allocated memory address when there's not enough space. When the memory stored in a row is not needed any more, the entire row shall be freed and set back to `NULL`.
+
+`TWODARRAY` is a kind of `THANDLE`, all of the `THANDLE`'s API apply to `TWODARRAY`. The following macros are provided with the same semantics as those of `THANDLE`'s:
+`TWODARRAY_INITIALIZE(T)`
+`TWODARRAY_ASSIGN(T)`
+`TWODARRAY_MOVE(T)`
+`TWODARRAY_INITIALIZE_MOVE(T)`
+
+
+## Exposed API
+
+```c
+
+/*to be used as the type of handle that wraps T - it has "rows", "cols" and "row_arrays"*/
+#define TWODARRAY(T)
+
+/*to be used in a header file*/
+#define TWODARRAY_TYPE_DECLARE(T)
+
+/*to be used in a .c file*/
+#define TWODARRAY_TYPE_DEFINE(T)
+```
+
+The macros expand to these useful somewhat more useful APIs:
+
+```c
+
+TWODARRAY(T) TWODARRAY_CREATE(T)(uint64_t row_size, uint64_t col_size);
+int TWODARRAY_FREE_ROW(T)(TWODARRAY(T) two_d_array, uint64_t row_index);
+int TWODARRAY_ALLOCATE_NEW_ROW(T)(TWODARRAY(T) two_d_array, uint64_t row_index);
+T* TWODARRAY_GET(T)(TWODARRAY(T) two_d_array, uint64_t row_index);
+void TWODARRAY_FREE(T)(TWODARRAY(T) two_d_array);
+
+```
+
+### TWODARRAY(T)
+
+```c
+#define TWODARRAY(T)
+```
+`TWODARRAY(T)` is a `THANDLE`(`TWODARRAY_STRUCT_T`), where `TWODARRAY_STRUCT_T` is a structure introduced like below:
+```c
```

Comment:
nitpick: you probably need to add a line between all these.

Reply:
added